src/utils/avatar-volume-utils.js

```diff
@@ -0,0 +1,56 @@
+export const MAX_VOLUME_LABELS = 21;
+export const MAX_GAIN_MULTIPLIER = 8;
+export const SMALL_GAIN_STEP = 1 / (MAX_VOLUME_LABELS / 2);
+export const SMALL_LEVEL_STEP = 1 / (MAX_GAIN_MULTIPLIER / 2);
+export const BIG_GAIN_STEP = (MAX_GAIN_MULTIPLIER - 1) / (MAX_VOLUME_LABELS / 2);
+export const BIG_LEVEL_STEP = 1;
+export const DEFAULT_VOLUME_BAR_MULTIPLIER = 1.0;
+
+export function calcGainStepUp(gainMultiplier) {
+  return gainMultiplier > 1 - SMALL_GAIN_STEP ? BIG_GAIN_STEP : SMALL_GAIN_STEP;
+}
+
+export function calcGainStepDown(gainMultiplier) {
+  return gainMultiplier > 1 + SMALL_GAIN_STEP ? BIG_GAIN_STEP : SMALL_GAIN_STEP;
+}
+
+export function calcLevelStepUp(level) {
+  return level > MAX_VOLUME_LABELS / 2 - SMALL_LEVEL_STEP ? BIG_LEVEL_STEP : SMALL_LEVEL_STEP;
+}
+
+export function calcLevelStepDown(level) {
+  return level > MAX_VOLUME_LABELS / 2 + SMALL_LEVEL_STEP ? BIG_LEVEL_STEP : SMALL_LEVEL_STEP;
+}
+
+export function calcLevel(gainMultiplier) {
+  return Math.min(
+    MAX_VOLUME_LABELS - 1,
+    gainMultiplier <= 1.001
+      ? Math.floor(gainMultiplier / SMALL_GAIN_STEP)
+      : Math.floor(MAX_VOLUME_LABELS / 2 + (gainMultiplier - 1) / BIG_GAIN_STEP)
+  );
+}
+
+export function calcGainMultiplier(level) {
+  return Math.min(
+    MAX_GAIN_MULTIPLIER,
+    level <= MAX_VOLUME_LABELS / 2 ? level * SMALL_GAIN_STEP : 1 + (level - MAX_VOLUME_LABELS / 2) * BIG_GAIN_STEP
+  );
+}
+
+export function updateAvatarVolumesPref(userId, gainMultiplier, muted) {
+  const avatarVoiceLevels = APP.store.state.preferences.avatarVoiceLevels || {};
+  avatarVoiceLevels[userId] = {
```

Comment:
I am actually kind of surprised this userId is consistent across sessions. I thought it was random every time you joined a room. If it's not we need to think about what that means privacy wise... But its kind of required to implement a feature like this...

Reply:
That name leaked from an initial implementation also, the right name for that parameter is be `displayName` as we don't really have access to any `userId` so we use the display name to store volumes. Updated.

Reply:
Hmm, displaynames are not guaranteed to be unique so that could pose a bit of an issue... I guess its probably the best we can do without making a decision on if we want to expose unique identifiers... Should be fairly easy to migrate later I guess. We should discuss this at the next engineering meeting.